src/rdkafka_int.h

```diff
@@ -339,6 +337,42 @@ struct rd_kafka_s {
                 /**< Lock for sparse_connect_random */
                 mtx_t         sparse_connect_lock;
         } rk_suppress;
+
+        /**
+         * SASL/OAUTHBEARER token value, token metadata, and extensions
+         */
+        struct {
+                /**< The lock protecting the below information */
+                rwlock_t refresh_lock;
+
+                /**< The b64token value as defined in RFC 6750 Section 2.1
+                 *   https://tools.ietf.org/html/rfc6750#section-2.1
+                 */
+                char *token_value;
+                /**< When the token expires, in terms of the number of
+                 *   milliseconds since the epoch.
+                 */
+                int64_t md_lifetime_ms;
+                /**< The point after which this token should be replaced with a
+                 * new one, in terms of the number of milliseconds since the
+                 * epoch.
+                 */
+                int64_t refresh_after_ms;
+                /**< When the last token refresh was equeued (0 = never)
+                 *   in terms of the number of milliseconds since the epoch.
+                 */
+                int64_t enqueued_refresh_ms;
+                /**< The name of the principal to which this token applies. */
+                char *md_principal_name;
+                /**< The SASL extensions, as per RFC 7628 Section 3.1
+                 *   https://tools.ietf.org/html/rfc7628#section-3.1
+                 */
+                rd_list_t extensions; /* rd_strtup_t list */
+                /**< Error message for validation and/or token retrieval problems.
+                 * Be sure to use snprintf(..errstr, sizeof(..errstr), ..)
+                 */
+                char errstr[512];
```

Comment:
Use an allocated pointer for errstr to avoid wasting memory.

Reply:
👍 Done